File: app/src/main/java/com/kaiserpudding/howtheywrite/lessonDetail/LessonDetailActivity.java
package com.kaiserpudding.howtheywrite.lessonDetail;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import com.kaiserpudding.howtheywrite.R;
import com.kaiserpudding.howtheywrite.model.Character;
import com.kaiserpudding.howtheywrite.model.Lesson;
import java.util.List;

public class LessonDetailActivity extends AppCompatActivity {

  public static String REPLY_LESSON_ID = "howTheyWrite.LESSON_ID";

  private LessonDetailViewModel lessonDetailViewModel;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_lesson_detail);

    RecyclerView recyclerView = findViewById(R.id.character_recyclerview);
    final LessonDetailAdapter adapter = new LessonDetailAdapter(this);
    recyclerView.setAdapter(adapter);
    recyclerView.setLayoutManager(new GridLayoutManager(this, 5));

    //TODO proper handling for 0
    int lessonId = getIntent().getIntExtra(REPLY_LESSON_ID, 0);

    //TODO proper fix for racing condition of initChar and getChar
    lessonDetailViewModel = ViewModelProviders.of(
        this, new LessonDetailViewModelFactory(getApplication(), lessonId))
        .get(LessonDetailViewModel.class);
    lessonDetailViewModel.getCharacters().observe(this, adapter::setCharacters);
  }
}

File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/lessonDetail/LessonDetailActivity.kt
package com.kaiserpudding.howtheywrite.lessonDetail

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.widget.Button
import com.kaiserpudding.howtheywrite.R
import com.kaiserpudding.howtheywrite.model.Character
import com.kaiserpudding.howtheywrite.quiz.QuizActivity

class LessonDetailActivity : AppCompatActivity() {

    private var lessonDetailViewModel: LessonDetailViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_lesson_detail)

        val recyclerView = findViewById<RecyclerView>(R.id.character_recyclerview)
        val adapter = LessonDetailAdapter(this)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = GridLayoutManager(this, 5)

        //TODO proper handling for 0
        val lessonId = intent.getIntExtra(REPLY_LESSON_ID, 0)

        //TODO proper fix for racing condition of initChar and getChar
        lessonDetailViewModel = ViewModelProviders.of(
                this, LessonDetailViewModelFactory(application, lessonId))
                .get(LessonDetailViewModel::class.java)
        lessonDetailViewModel!!.characters.observe(this, Observer<List<Character>> { adapter.setCharacters(it!!) })

        val toQuiz = findViewById<Button>(R.id.to_quiz_button)
        toQuiz.setOnClickListener { view ->
            val intent = Intent(view.context, QuizActivity::class.java)
            intent.putExtra(REPLY_LESSON_ID, lessonId)
            startActivity(intent)
        }

    }

    companion object {

        var REPLY_LESSON_ID = "howTheyWrite.LESSON_ID"
    }
}
